### DISCUSSION TITLE
[Question]: Force Node.js to crash if the `end` event is emitted by the Redis client

### DISCUSSION THREAD

**Discussion by LebonNic** [taskforcesh/bullmq#1814] [Q&A]
Hello,

First of all, thanks for the work you are doing on this lib.

I have a question because I can't find any information in the documentation about this case. Using the following code to instanciate a `Queue` or `Worker` (so letting `BullMQ` create the `Redis` client):

```javascript
const Worker = require("bullmq").Worker;
const MetricsTime = require("bullmq").MetricsTime;

createWoker(queueName) {
    const config = {
      port: process.env.REDIS_PORT,
      host: process.env.REDIS_HOST,
      password: process.env.REDIS_PASSWORD,
    };
    
    const self = this;
    const worker = new Worker(
      queueName,
      async (job) => {
        await self.processJob(job);
      },
      {
        connection: config,
        metrics: {
          maxDataPoints: MetricsTime.ONE_WEEK,
        },
      }
    );
    
    return worker;
  }
```

What happens if the `end` event is issued by the `Redis` client? The IORedis [documentation](https://github.com/luin/ioredis#connection-events) states that :

> `end` emits after `close` when no more reconnections will be made, or the connection is failed to establish.

In other words, the maximum number of attempts to reconnect to `Redis` is exceeded and the connection is lost. In this case, the node on which `BullMQ` is running is no longer useful and should be stopped. So I wanted to know if there was a mechanism to force the node to crash? Like in this example:

```javascript
const Redis = require("ioredis");

buildRedisCient(config) {
    const self = this;
    const client = new Redis(config);

    client.on("end", (args) => {
      self.logger.error(`End event received: stop the runner! (args: ${args})`);
      process.exit(1);
    });

    return client;
  }
```

In this way, it may be my particular case, but my cloud provider is able to automatically restart the nodes when these have been killed and most of the time, when redeploying the application, the new nodes reconnect perfectly to the Redis database.

Thank you in advance,
Nicolas

**Comment by manast** [comment]
I wonder if this is something that can happen in practice. As long as BullMQ instantiates the Redis connection, the "end" event would only be triggered when closing the Worker or Queue that instantiated said Redis connection. On the other hand, if the connection is instantiated outside of BullMQ, then yes, the "end" event is ignored, but basically, you should not close the Redis connection before you close the Queue or Workers instances to avoid issues. 
Here some pieces of documentation that I think are relevant for this question: 
https://docs.bullmq.io/patterns/failing-fast-when-redis-is-down
https://docs.bullmq.io/guide/going-to-production#automatic-reconnections
